plaidcloud/utilities/frame_manager.py

```diff
@@ -2434,43 +2460,73 @@ def excel_to_csv_xlrd(excel_file_name, csv_file_name, sheet_name='sheet1', clean
                         # Skip rows that have no data.
-                        skipped_rows += 1
                         continue
-                wr.writerow([
-                    (
-                        c.value
-                        if c.ctype not in [xlrd.XL_CELL_DATE, xlrd.XL_CELL_EMPTY, xlrd.XL_CELL_BLANK, xlrd.XL_CELL_ERROR, xlrd.XL_CELL_NUMBER]
-                        # else xlrd.error_text_from_code[c.value]  # if you wanted the error text instead of NULL
-                        # if c.ctype == xlrd.XL_CELL_ERROR
-                        else '<NULL>'
-                        if c.ctype in [xlrd.XL_CELL_EMPTY, xlrd.XL_CELL_BLANK, xlrd.XL_CELL_ERROR]
-                        else int(c.value)
-                        if c.ctype == xlrd.XL_CELL_NUMBER and c.value == int(c.value)
-                        else get_formatted_number(c.value)
-                        if c.ctype == xlrd.XL_CELL_NUMBER
-                        else datetime.datetime(
-                            *xlrd.xldate_as_tuple(c.value, datemode)
-                        ).isoformat()
-                        if xlrd.xldate_as_tuple(c.value, datemode)[0] != 0
-                        else datetime.time(
-                            *xlrd.xldate_as_tuple(c.value, datemode)[:3]
-                        ).isoformat()
-                    )
-                    for c in sh.row(rownum)
-                ])
+
+                row = []
+                col_pos = 0
+                for col_info in column_information:
+
+                    c = sh.cell(rownum, col_pos)
+
+                    if c.ctype in (xlrd.XL_CELL_EMPTY, xlrd.XL_CELL_BLANK):
+                        row.append(null_value)
+                    else:
+                        dtype = dtype_from_excel(c.ctype)
+                        # logger.info(f'Column: {col_pos}')
+                        if dtype != col_info['dtype']:
+                            column_information[col_pos]['dtype'] = 'text' # elegantly handle situation where the guess was wrong initially.  must be mixed data.  default to text.
+
+                        if dtype == 'text':
+                            row.append(c.value)
+                        elif dtype == 'numeric':
+                            if int(c.value) ==  c.value:
+                                row.append(int(c.value)) # This appears to be an int
+                            else:
+                                row.append(get_formatted_number(c.value)) # some other type of number
+                        elif dtype == 'boolean':
+                            row.append(c.value)
+                        elif dtype == 'datetime':
+                            if xlrd.xldate_as_tuple(c.value, datemode)[0] == 0:
+                                row.append(null_value)
+                            else:
+                                row.append(datetime.time(*xlrd.xldate_as_tuple(c.value, datemode)[:3]).isoformat())
```

Comment:
This looks setup the wrong way around. Should prob read something like:
```
                            date_tuple = xlrd.xldate_as_tuple(c.value, datemode)
                            if date_tuple[0] == 0:
                                row.append(datetime.time(*date_tuple[:3]).isoformat())
                            else:
                                row.append(datetime.datetime(*date_tuple).isoformat())
```

Reply:
Actually, looking back up in the code, there is a `xlrd.xldate_as_datetime` method